sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveCompat.scala

```diff
@@ -0,0 +1,182 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.hive
+
+import java.math.{BigDecimal => JBigDecimal}
+import java.util.{Properties, Set => JSet}
+
+import scala.collection.JavaConversions._
+import scala.language.{existentials, implicitConversions}
+
+import org.apache.hadoop.conf.Configuration
+import org.apache.hadoop.fs.Path
+import org.apache.hadoop.hive.common.`type`.HiveDecimal
+import org.apache.hadoop.hive.conf.HiveConf
+import org.apache.hadoop.hive.ql.metadata.{Hive, Partition, Table}
+import org.apache.hadoop.hive.ql.plan.{FileSinkDesc, TableDesc}
+import org.apache.hadoop.hive.ql.processors.{CommandProcessor, CommandProcessorFactory}
+import org.apache.hadoop.hive.serde2.{ColumnProjectionUtils, Deserializer}
+import org.apache.hadoop.mapred.InputFormat
+
+import org.apache.spark.Logging
+
+/**
+ * A utility object used to cope with Hive compatibility issues.
+ */
+object HiveCompat {
+  def createDefaultDBIfNeeded(context: HiveContext) = {
+    context.runSqlHive("CREATE DATABASE IF NOT EXISTS default")
+    context.runSqlHive("USE default")
+  }
+
+  def newTableDesc(
+      serdeClass: Class[_ <: Deserializer],
+      inputFormatClass: Class[_ <: InputFormat[_, _]],
+      outputFormatClass: Class[_],
+      properties: Properties) = callWithAlternatives(
+    // For Hive 0.13.1
+    Construct(classOf[TableDesc],
+      classOf[Class[_ <: InputFormat[_, _]]] -> inputFormatClass,
+      classOf[Class[_]] -> outputFormatClass,
+      classOf[Properties] -> properties),
+
+    // For Hive 0.12.0
+    Construct(classOf[TableDesc],
+      classOf[Class[_ <: Deserializer]] -> serdeClass,
+      classOf[Class[_ <: InputFormat[_, _]]] -> inputFormatClass,
+      classOf[Class[_]] -> outputFormatClass,
+      classOf[Properties] -> properties))
+
+  def getCommandProcessor(cmd: Array[String], conf: HiveConf) = {
+    callWithAlternatives[CommandProcessor](
+      // For Hive 0.13.1
+      InvokeStatic[CommandProcessor](classOf[CommandProcessorFactory], "get",
+        classOf[Array[String]] -> cmd,
+        classOf[HiveConf] -> conf),
+
+      // For Hive 0.12.0
+      InvokeStatic[CommandProcessor](classOf[CommandProcessorFactory], "get",
+        classOf[String] -> cmd(0),
+        classOf[HiveConf] -> conf))
+  }
+
+  def getAllPartitionsOf(client: Hive, tbl: Table) = callWithAlternatives[JSet[Partition]](
+    // For Hive 0.13.1
+    Invoke[JSet[Partition]](classOf[Hive], client, "getAllPartitionsOf",
+      classOf[Table] -> tbl),
+
+    // For Hive 0.12.0
+    Invoke[JSet[Partition]](classOf[Hive], client, "getAllPartitionsForPruner",
+      classOf[Table] -> tbl)
+  )
+
+  def newHiveDecimal(bd: JBigDecimal) = callWithAlternatives(
```

Comment:
I probably shouldn't reimplement this one via reflection, as it can be performance sensitive.